shell/platform/fuchsia/flutter/software_surface_producer.cc

```diff
@@ -0,0 +1,236 @@
+    const SkISize& size) {
+  TRACE_EVENT2("flutter", "SoftwareSurfacePool::ProduceSurface", "width",
+               size.width(), "height", size.height());
+  FML_CHECK(valid_);
+
+  std::unique_ptr<SurfaceProducerSurface> surface;
+  auto exact_match_it =
+      std::find_if(available_surfaces_.begin(), available_surfaces_.end(),
+                   [&size](const auto& surface) {
+                     return surface->IsValid() && surface->GetSize() == size;
+                   });
+  if (exact_match_it != available_surfaces_.end()) {
+    TRACE_EVENT_INSTANT0("flutter", "Exact match found");
+    surface = std::move(*exact_match_it);
+    available_surfaces_.erase(exact_match_it);
+  } else {
+    surface = CreateSurface(size);
+  }
+
+  if (surface == nullptr) {
+    FML_LOG(ERROR) << "SoftwareSurfaceProducer: Could not acquire surface";
+    return nullptr;
+  }
+
+  if (!surface->FlushSessionAcquireAndReleaseEvents()) {
+    FML_LOG(ERROR)
+        << "SoftwareSurfaceProducer: Could not flush acquire/release "
+           "events for buffer.";
+    return nullptr;
+  }
+
+  return surface;
+}
+
+void SoftwareSurfaceProducer::SubmitSurfaces(
+    std::vector<std::unique_ptr<SurfaceProducerSurface>> surfaces) {
+  TRACE_EVENT0("flutter", "SoftwareSurfaceProducer::SubmitSurfaces");
+
+  // Submit surface
+  for (auto& surface : surfaces) {
+    SubmitSurface(std::move(surface));
+  }
+
+  // Buffer management.
+  AgeAndCollectOldBuffers();
+}
+
+void SoftwareSurfaceProducer::SubmitSurface(
+    std::unique_ptr<SurfaceProducerSurface> surface) {
+  TRACE_EVENT0("flutter", "SoftwareSurfacePool::SubmitSurface");
+  FML_CHECK(valid_);
+
+  // This cast is safe because |SoftwareSurface| is the only implementation of
+  // |SurfaceProducerSurface| for Flutter on Fuchsia.  Additionally, it is
+  // required, because we need to access |SoftwareSurface| specific information
+  // of the surface (such as the amount of memory it contains).
+  auto software_surface = std::unique_ptr<SoftwareSurface>(
+      static_cast<SoftwareSurface*>(surface.release()));
+  if (!software_surface) {
+    return;
+  }
+
+  uintptr_t surface_key = reinterpret_cast<uintptr_t>(software_surface.get());
+  auto insert_iterator = pending_surfaces_.insert(std::make_pair(
+      surface_key,                 // key
+      std::move(software_surface)  // value
+      ));
+  if (insert_iterator.second) {
+    insert_iterator.first->second->SignalWritesFinished(std::bind(
+        &SoftwareSurfaceProducer::RecyclePendingSurface, this, surface_key));
+  }
+}
+
+std::unique_ptr<SoftwareSurface> SoftwareSurfaceProducer::CreateSurface(
+    const SkISize& size) {
+  TRACE_EVENT2("flutter", "SoftwareSurfacePool::CreateSurface", "width",
+               size.width(), "height", size.height());
+  auto surface =
+      std::make_unique<SoftwareSurface>(sysmem_allocator_, flatland_allocator_,
+                                        scenic_session_, size, buffer_id_++);
```

Comment:
Should we use a variant here between flatland and scenic?

Reply:
+1, I feel like this would make the code easier to follow even if it's fairly redundant between the two. The amount of FML_CHECKs/functions that don't do anything based on the presence of a session make me feel like the class will be difficult to reason about.

Reply:
I agree that it makes the code much harder to reason about.

The thing is - The scenic dependency here is a false dependency; this code doesn't actually *need* a dependency on scenic and it is only here due to smelly legacy code.

See https://bugs.fuchsia.dev/p/fuchsia/issues/detail?id=88466 for a fuller description of what is going on and how to fix this.

Since a variant would encourage deeper intertwining of the code, and the goal is to break the dependency, I don't want to use one here.  Instead, we can reduce the code complexity by fixing 88466.